### ISSUE TITLE
Widget screenshot in `magicgui.md` error

### ISSUE THREAD

**Issue by lucyleeow** [napari/docs#283]
## 🐛 Bug

Widget screenshot in `magicgui.md` (in dev docs: https://napari.org/dev/guides/magicgui.html) looks wrong:

![image](https://github.com/napari/docs/assets/23182829/badccb87-75f2-431e-881a-eaf093d6f397)

Not familiar with `magicgui` so not sure if this is due to our example or the screenshot tool?

**Comment by Czaki** [comment]
Could you put a link to webpage? Or it happens when build locally?

**Comment by psobolewskiPhD** [comment]
It's here, in the dev docs:
https://napari.org/dev/guides/magicgui.html
Stable looks file.
Screenshot below also looks fine, so maybe it's just some glitch? Let me check a CircleCI preview.
Edit: Here's the same page on a CircleCI preview from a most recently updated PR:
https://output.circle-artifacts.com/output/job/6ed6c9c2-7037-4f5c-84e0-93d5c4d1472f/artifacts/0/docs/docs/_build/guides/magicgui.html
Looks fine, so I think it will be fixed by the next commit to main. Going to leave this open so we remember to double check it.

**Comment by lucyleeow** [comment]
Interesting. Maybe it's a browser problem, I'll check with chromium

**Comment by lucyleeow** [comment]
Whoops mis read. It looks fine on CI for me too so hopefully fixed by commit

**Comment by Czaki** [comment]
I have merged new commit and problem still exists (image is regenerated).

**Comment by psobolewskiPhD** [comment]
I think it would have to be merged to napari/docs/main?

**Comment by Czaki** [comment]
the dev docs are generated on every merge to main of napari/napari repository.

**Comment by lucyleeow** [comment]
> Screenshot below also looks fine,

Which screenshot? Just to clarify you see the same problem in dev?

(dev vs stable)

Do we generate stable/dev differently? Maybe a different magicgui version (is causing the problem)?

**Comment by Czaki** [comment]
We have not regenerated stable since the 0.4.18 release.

**Comment by psobolewskiPhD** [comment]
> > Screenshot below also looks fine,
> 
> Which screenshot? Just to clarify you see the same problem in dev?

The napari generated screenshot just below the wierd one.
Also whats with the `add caption here`?

**Comment by lucyleeow** [comment]
> Also whats with the add caption here?

I know right?! I was very confused by this :joy: 

> We have not regenerated stable since the 0.4.18 release.

The example does not use napari at all. Just `magicgui`, `datetime` and `pathlib` (thus maybe change to `magicgui` version or screenshotting tool??)

**Comment by Czaki** [comment]
> Also whats with the `add caption here`?

https://github.com/napari/docs/blob/de3c3e00f80584710c72984328e7835b3793132a/docs/guides/magicgui.md?plain=1#L51

It is here in code. Some bug in during code review happens.

**Comment by lucyleeow** [comment]
Oh yeah I knew it was in the code. I did not know the intention but yes probably missed in code review.

**Comment by Czaki** [comment]
I have found that we have two workflows for build docs:

https://github.com/napari/docs/blob/main/.github/workflows/deploy_docs.yml
and
https://github.com/napari/docs/blob/main/.github/workflows/build_docs.yml

The deploy one do not use constraints in `make` steep. 

I think that we should use a single workflow with conditional deeploy steep.

**Comment by lucyleeow** [comment]
Does the code used:

```python
from magicgui import magicgui
import datetime
import pathlib

@magicgui(
    call_button="Calculate",
    slider_float={"widget_type": "FloatSlider", 'max': 10},
    dropdown={"choices": ['first', 'second', 'third']},
)
def widget_demo(
    maybe: bool,
    some_int: int,
    spin_float=3.14159,
    slider_float=4.5,
    string="Text goes here",
    dropdown='first',
    date=datetime.datetime.now(),
    filename=pathlib.Path('/some/path.ext')
):
    ...

widget_demo.show()
```

work for others? It generates an empty widget for me (but I am likely to be running it wrong).

**Comment by Czaki** [comment]
You need to launch event loop to have visible widget:

![obraz](https://github.com/napari/docs/assets/3826210/29ea2568-f2a1-41ab-a933-e1e949316faf)


```python
from qtpy.QtWidgets import QApplication

from magicgui import magicgui
import datetime
import pathlib

app = QApplication([])

@magicgui(
    call_button="Calculate",
    slider_float={"widget_type": "FloatSlider", 'max': 10},
    dropdown={"choices": ['first', 'second', 'third']},
)
def widget_demo(
    maybe: bool,
    some_int: int,
    spin_float=3.14159,
    slider_float=4.5,
    string="Text goes here",
    dropdown='first',
    date=datetime.datetime.now(),
    filename=pathlib.Path('/some/path.ext')
):
    ...

widget_demo.show()
app.exec()
```

**Comment by lucyleeow** [comment]
Ahhhh okay, I see we have:

https://github.com/napari/docs/blob/de3c3e00f80584710c72984328e7835b3793132a/docs/conf.py#L219

which does the app stuff.

**Comment by Czaki** [comment]
But this iterates over napari main windowses. How should it discover the magicgui widget?

**Comment by lucyleeow** [comment]
I see we have:

https://github.com/napari/docs/blob/de3c3e00f80584710c72984328e7835b3793132a/.github/workflows/build_docs.yml#L1-L2

so there may be a reason we didn't/couldn't use the same workflow, @melissawm ?

Deploy uses constraints:

https://github.com/napari/docs/blob/de3c3e00f80584710c72984328e7835b3793132a/.github/workflows/deploy_docs.yml#L49-L50

Build (used in PRs), just uses the `requirements.txt` file:

https://github.com/napari/docs/blob/de3c3e00f80584710c72984328e7835b3793132a/.github/workflows/build_docs.yml#L46

**Comment by lucyleeow** [comment]
> But this iterates over napari main windowses. How should it discover the magicgui widget?

I am not good with Qt stuff but maybe this will answer your question: https://sphinx-gallery.github.io/stable/advanced.html#write-a-custom-image-scraper

tl;dr for each 'code' block/section in the docs, images are scraped using `napari_scraper`. I think the code in the code block is executed and then the scraper code (is run to get any images resulting from the executing the code block).

**Comment by Czaki** [comment]
> so there may be a reason we didn't/couldn't use the same workflow, @melissawm ?

I'm author of this note. It is because when I was fixing build_docs workflows I miss deploy docs one. 

Build also uses constraints but pass them using enviornment variable:

https://github.com/napari/docs/blob/de3c3e00f80584710c72984328e7835b3793132a/.github/workflows/build_docs.yml#L46-L48

Also make uses constraints 
https://github.com/napari/docs/blob/de3c3e00f80584710c72984328e7835b3793132a/.github/workflows/build_docs.yml#L57-L63

When deploy do not use this:

https://github.com/napari/docs/blob/de3c3e00f80584710c72984328e7835b3793132a/.github/workflows/deploy_docs.yml#L59-L66

**Comment by lucyleeow** [comment]
Thanks but I do not follow what `PIP_CONSTRAINT` does here:

https://github.com/napari/docs/blob/de3c3e00f80584710c72984328e7835b3793132a/.github/workflows/build_docs.yml#L57-L60

since in "Install Dependencies" step above we used the constraints?

> I think that we should use a single workflow with conditional deploy steep.

I'll make another issue to see if we can use the same workflow for both.

**Comment by lucyleeow** [comment]
Ahhh i see we changed the way we scraped images right after 0.4.18 release: https://github.com/napari/docs/pull/207

@aganders3 could this be why the `magicgui` widget does not work? I guess since this scraper is specific for napari?

**Comment by aganders3** [comment]
Eek I'm just catching up on this but yes, it's a possibility. I can take a look at this tomorrow.

**Comment by psobolewskiPhD** [comment]
We use that scraper in the CircleCI docs which look fine so I dont think that's it. Note they use the update workflow like build_docs.

**Comment by lucyleeow** [comment]
> update workflow

What is the update workflow?

What is the difference in the environment between the "build docs" and "build and deploy"? So I can try to replicate locally. To me they both seem to use the same constraints..?

https://github.com/napari/docs/issues/283#issuecomment-1837628576

**Comment by psobolewskiPhD** [comment]
Sorry I wasn't clear, this PR updated the build_docs and the circle ci workflows, but not the deploy one:
https://github.com/napari/docs/pull/266
The issue is that the `make docs` step has a `pip install` step in it:
https://github.com/napari/docs/blob/de3c3e00f80584710c72984328e7835b3793132a/Makefile#L24
So without the env variable, that can install different versions of stuff than  the previous steps with constraints, which is what Grzegorz was pointing out.

**Comment by lucyleeow** [comment]
Ahhh I did not think the Makefile would have an install step!

This setup seems odd? Why do we need the "Install Dependencies" step in `build_docs` then? Just to print napari version and `napari.layers.__doc__` ?

**Comment by psobolewskiPhD** [comment]
I think some of this is vestigial? The idea was probably a simple `make docs` and when we have mono repo this probably worked. Now we need to install napari and then install the docs dependencies. Plus we have the constraints making sure nothing is broken by new versions of dependencies.
I think for the CI workflows, we could also just switch from `make docs`
https://github.com/napari/docs/blob/de3c3e00f80584710c72984328e7835b3793132a/Makefile#L35
 to `make clean docs-build` and skip the extra install step inside the makefile.

**Comment by aganders3** [comment]
I'm just poking around this morning. Docs look okay for me when built locally, but I will say I am a little confused still how that first magicgui (non-napari) image works at all.

> Also whats with the `add caption here`?

This might be a red herring as it's actually just...in the doc?
https://github.com/napari/docs/blob/de3c3e00f80584710c72984328e7835b3793132a/docs/guides/magicgui.md?plain=1#L51

**Comment by aganders3** [comment]
Okay now I am understanding more. This has nothing to do with the scraper because that is for the gallery only. These screenshots come from [myst-nb](https://myst-nb.readthedocs.io/), which renders the output of the cells in that file because they use the [`code-cell` directive](https://myst-nb.readthedocs.io/en/latest/authoring/basics.html#text-based-notebooks).

The first widget shows in the notebook because it provides [`_repr_png_`](https://github.com/pyapp-kit/magicgui/blob/33c5e60ab02d5dc1e9b31e23e62303e0ae555bbe/src/magicgui/widgets/bases/_widget.py#L390).

The rest of the widgets are rendered more manually by combining two code blocks: one with `remove-output` (shows the code) and one with `remove-input` (shows a screenshot):
https://github.com/napari/docs/blob/de3c3e00f80584710c72984328e7835b3793132a/docs/guides/magicgui.md?plain=1#L110-L131

It looks like maybe Qt was in some weird state here. Each screenshot sets its own size but in the output they successively shrink for some reason. edit: maybe adding some `viewer.close()` or `napari.Viewer.close_all()` (perhaps even in other files) would help?

**Comment by melissawm** [comment]
> I'm just poking around this morning. Docs look okay for me when built locally, but I will say I am a little confused still how that first magicgui (non-napari) image works at all.
> 
> > Also whats with the `add caption here`?
> 
> This might be a red herring as it's actually just...in the doc?
> 
> https://github.com/napari/docs/blob/de3c3e00f80584710c72984328e7835b3793132a/docs/guides/magicgui.md?plain=1#L51

We had a sprint for alt-text once and marked this for someone to write something since IIRC the screenshot for magicgui does not support alt-text out of the box. If someone can add a description that would be great 😄

**Comment by melissawm** [comment]
> I think some of this is vestigial? The idea was probably a simple `make docs` and when we have mono repo this probably worked. Now we need to install napari and then install the docs dependencies. Plus we have the constraints making sure nothing is broken by new versions of dependencies. I think for the CI workflows, we could also just switch from `make docs`
> 
> https://github.com/napari/docs/blob/de3c3e00f80584710c72984328e7835b3793132a/Makefile#L35
> 
> to `make clean docs-build` and skip the extra install step inside the makefile.

To be honest, ideally I would love for the dependency install to be de-coupled from the docs build command. I understand the convenience but I feel like it's confusing to have them together. But that's a discussion for another time 😅

**Comment by melissawm** [comment]
> I see we have:
> 
> https://github.com/napari/docs/blob/de3c3e00f80584710c72984328e7835b3793132a/.github/workflows/build_docs.yml#L1-L2
> 
> so there may be a reason we didn't/couldn't use the same workflow, @melissawm ?
> 
> Deploy uses constraints:
> 
> https://github.com/napari/docs/blob/de3c3e00f80584710c72984328e7835b3793132a/.github/workflows/deploy_docs.yml#L49-L50
> 
> Build (used in PRs), just uses the `requirements.txt` file:
> 
> https://github.com/napari/docs/blob/de3c3e00f80584710c72984328e7835b3793132a/.github/workflows/build_docs.yml#L46

The reason is that those two workflows were similar but one was updated and the other wasn't. So +1 to getting deploy_docs to be the same as build_docs!

**Comment by lucyleeow** [comment]
> To be honest, ideally I would love for the dependency install to be de-coupled from the docs build command.

+1 here. I think it would be great if using `make docs` did not install things, as it may cause unintended consequences for the user if they did not know it included installation. I am not that experienced with docs but but I've never seen `make docs`  install packages.

(I'll start an issue anyway so we can have a discussion)

**Comment by lucyleeow** [comment]
> maybe adding some viewer.close() or napari.Viewer.close_all() (perhaps even in other files) would help?

Yes it is probably safer to do something like this. The annoying part though is that we haven't been able to replicate this problem locally. I guess we will need to try installing from requirements as is/was in `deploy_docs` (via the makefile)

Edit: of note, the docs look fine in CI, so it's not a change in napari that has cause this.

**Comment by lucyleeow** [comment]
@melissawm do you know if there is a reason we use sphinx-gallery for code outputs in the gallery but myst-nb elsewhere in the doc? It would be nice to consistently use one tool?

Edit: I guess it would involve a lot of re-arranging to use sphinx gallery for non-gallery docs, especially as they use RST. Is myst-nb an option for gallery docs?

**Comment by psobolewskiPhD** [comment]
So I'm pretty sure that the deploy workflow worked properly now that #290 was merged.
The deployment action finished on the GitHub.io repo.
And yet, the widget is still messed up like before:
https://napari.org/dev/guides/magicgui.html
But the preview docs on CircleCI look fine:
https://output.circle-artifacts.com/output/job/a6aac828-d441-45bc-871d-c0ae279fd1f2/artifacts/0/docs/docs/_build/guides/magicgui.html
Also, the zip artifact (GH runners build_docs) for #290 (https://github.com/napari/docs/actions/runs/7093485902)
has the same issue. But when I build locally, the image is fine.
So it seems like it's related to the GH runners? 🤔

**Comment by psobolewskiPhD** [comment]
I spot two differences between GH and CircleCI (and my local build): we use Talleys QT action:
https://github.com/napari/docs/blob/0c7029e5c9674e9d56c5efbec1022fa302661280/.github/workflows/build_docs.yml#L40
This action seems to run:  
https://github.com/tlambert03/setup-qt-libs/blob/b66de97cf84c0684fa084b8487cfd4ae2c622f40/dist/index.js#L2982

While in CircleCI we install manually:
https://github.com/napari/docs/blob/0c7029e5c9674e9d56c5efbec1022fa302661280/.circleci/config.yml#L22-L23
The difference appears that the Talley action also has `libxcb-cursor0 libopengl0 libegl1-mesa` while circleci has `xvfb`.
But looking at the log it looks like xvfb does pull `libegl-mesa0` (note the 0, not sure it's important) but not the others :
https://app.circleci.com/pipelines/github/napari/docs/413/workflows/b7d5ed05-1315-4de4-8689-9ed1ee574100/jobs/403?invite=true#step-103-3065_45
(And locally obviously I have macOS stuff)
Then in GH workflow we use the @aganders3 headless run:
https://github.com/napari/docs/blob/0c7029e5c9674e9d56c5efbec1022fa302661280/.github/workflows/build_docs.yml#L56
while CircleCI uses xvfb:
https://github.com/napari/docs/blob/0c7029e5c9674e9d56c5efbec1022fa302661280/.circleci/config.yml#L43
and locally I get the windows dancing.

**Comment by aganders3** [comment]
> Then in GH workflow we use the @aganders3 headless run

I thought I was off the hook when we determined it wasn't the scraper 😆.

Buuuut - this action runs a tiling window manager by default, so maybe this could cause the shrinking screenshots. I don't really know how it would mess up that initial screenshot but who knows. So my previous suggestion of using `viewer.close()` might also work here. Alternatively the action can be configured by setting `linux-pkgs` and `linux-setup`, so maybe setting something like this would work:

```diff
      - name: Build Docs
        uses: aganders3/headless-gui@v1
        env:
          GOOGLE_CALENDAR_ID: ${{ secrets.GOOGLE_CALENDAR_ID }}
          GOOGLE_CALENDAR_API_KEY: ${{ secrets.GOOGLE_CALENDAR_API_KEY }}
          PIP_CONSTRAINT: ${{ github.workspace }}/napari/resources/constraints/constraints_py3.10_docs.txt
        with:
          run:  make -C docs docs
+          linux-pkgs: ""
+          linux-setup: ""
```

**Comment by aganders3** [comment]
I may be reading it wrong but it also looks like the CircleCI build is using PySide and the GH-Actions build is using PyQt5 (from `napari[all]`).

**Comment by psobolewskiPhD** [comment]
Good catch. I'm using pyqt6 locally, which is also fine, so a glitch in pyqt5 specifically is also possible?

**Comment by aganders3** [comment]
Could still be specific to headless too. Also I can see there's a different code path for this rendering code in magicgui depending on PySide/PyQt. I'm otherwise running low on ideas 😬.

**Comment by aganders3** [comment]
I tried PySide in my PR and get _no_ image now but an exception instead (in the docs output):

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File /opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/IPython/core/formatters.py:344, in BaseFormatter.__call__(self, obj)
    342     method = get_real_method(obj, self.print_method)
    343     if method is not None:
--> 344         return method()
    345     return None
    346 else:

File /opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/magicgui/widgets/bases/_widget.py:403, in Widget._repr_png_(self)
    397     print(
    398         "(For a nicer magicgui widget representation in "
    399         "Jupyter, please `pip install imageio`)"
    400     )
    401     return None
--> 403 rendered = self.render()
    404 if rendered is not None:
    405     with BytesIO() as file_obj:

File /opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/magicgui/widgets/bases/_widget.py:384, in Widget.render(self)
    382 def render(self) -> np.ndarray:
    383     """Return an RGBA (MxNx4) numpy array bitmap of the rendered widget."""
--> 384     return self._widget._mgui_render()

File /opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/magicgui/backends/_qtpy/widgets.py:187, in QBaseWidget._mgui_render(self)
    185 h, w, c = img.height(), img.width(), 4
    186 if qtpy.API_NAME.startswith("PySide"):
--> 187     arr = np.array(bits).reshape(h, w, c)
    188 else:
    189     bits.setsize(h * w * c)

ValueError: cannot reshape array of size 204160 into shape (290,352,4)
```

Here's the run: https://github.com/napari/docs/actions/runs/7108661085

**Comment by aganders3** [comment]
Okay I have another theory now that this is related to the bit-depth of the xvfb (virtual) screen.

In my `headless-gui` action the screen is set here:
https://github.com/aganders3/headless-gui/blob/5124e2e44f3012512071e525269b50d0cadc3ef7/src/start-xvfb.bash#L25-L26

In CircleCI it just uses `xvfb-run` with defaults, which I ~~think is `640x480x8`~~ don't know.

I will see if I can figure out a reasonable way to test this theory tomorrow 😓.

**Comment by psobolewskiPhD** [comment]
Ooo! Bit depth difference could maybe explain the funky colors?
https://manpages.ubuntu.com/manpages/xenial/man1/xvfb-run.1.html
suggests the default is `-screen 0 640x480x8`

**Comment by psobolewskiPhD** [comment]
Fixed by https://github.com/napari/docs/pull/294
